practice/rain.py

```diff
@@ -1,86 +1,214 @@
+    [['Charmander Bulbasuar Squirtle'], ['Ponya Oddish Magicarp'], ['Fire Plant Water']]
     """
     rows = [split_single_line_to_single_row(line) for line in text]
     return rows
 
 
 def convert_rows_to_columns(rows):
     """Takes rows as lists of strings, and transforms them into columns as lists of strings
-    >>> [['Charmander', 'Bulbasuar', 'Squirtle'],['Ponya', 'Oddish', 'Magicarp'],['Fire', 'Plant', 'Water']]
-    [['Charmander', 'Ponya'', 'Fire'],['Bulbasuar', 'Oddish', 'Plant'],['Squirtle', 'Magicarp', 'Water']]
+
+    >>> convert_rows_to_columns([['Char', 'Bulb', 'Squirt'],['Ponya', 'Oddish', 'Magicarp'],['Fire', 'Plant', 'Water']])
+    [['Char', 'Ponya', 'Fire'], ['Bulb', 'Oddish', 'Plant'], ['Squirt', 'Magicarp', 'Water']]
     """
-    columns = list(zip(*rows))
+    columns_by_tuple = list(zip(*rows))
+    columns = [list(tup) for tup in columns_by_tuple]
     return columns
 
 
-def get_most_rain_inches(columns):
+def contains_numeral(value):
+    """Filtering function, returns boolean if string is numeric.
+    >>> contains_numeral('13')
+    True
+    >>> contains_numeral('pie')
+    False
+    """
+    try:
+        int(value)
+        return True
+    except ValueError:
+        return False
+
+
+def get_most_daily_rain_inches(columns):
     """Gets the value of the most rain for any given date
-    >>> get_most_rain_inches([['1', '2', '3', '8'], ['4', '5', '6', '1'], ['7', '8', '9', '12']])
-    '.06'
+
+    >>> get_most_daily_rain_inches([['1', '2', '3', '8'], ['4', '5', '6', '1'], ['7', '8', '9', '12'], \
+    ['4', '7', '6', '1']])
+    '0.07'
     """
-    most_rain = max(columns[1])
+    rainy_days = [int(value) for value in columns[3] if contains_numeral(value)]
+    most_rain = max(rainy_days)
     most_rain_inches = str(int(most_rain) / 100)
     return most_rain_inches
 
 
 def get_index_of_max_daily_rain_box(columns):
     """Takes in rows of data, and returns index of day with most rain columns max rain
-    >>> get_day_with_most_rain([['1', '2', '3', '8'], ['4', '5', '6', '1'], ['7', '8', '9', '12']])
-    '2'
+
+    >>> get_index_of_max_daily_rain_box(TEST_COLUMNS)
+    0
     """
-    index_of_max_daily_rain = columns[1].index(max(columns[1]))
-    return index_of_max_daily_rain
+    rainy_days = [int(value) for value in columns[3] if contains_numeral(value)]
+    daily_rain_values = [value for value in rainy_days]
+    day_index_of_max_daily_rain = daily_rain_values.index(max(daily_rain_values))
+    return day_index_of_max_daily_rain
 
 
-def get_day_with_most_rain (rows, columns):
+def get_day_with_most_rain(rows, columns):
+    """Returns the date of day with the most rain
+
+    >>> get_day_with_most_rain(TEST_ROWS, TEST_COLUMNS)
+    'MAR 23 2016'
+    """
     index_of_day_cell = get_index_of_max_daily_rain_box(columns)
     day_full_line = rows[index_of_day_cell]
-    date = day_full_line[0]
+    date = day_full_line[1] + ' ' + day_full_line[0] + ' ' + day_full_line[2]
     return date
```

Comment:
A general suggestion I have for structuring data: keep pairs paired. This algorithm, spread out over the functions above, effectively:

1. Slices out all day measurements.
1. Sums them.
1. Figures out max.
1. Looks back through all of the sums for an index.
1. Does a coordinated lookup at the same index for the date.

This works! But it's fragile. You have 3? lists that you have to keep in exactly the same order to ensure that you get the right answer by doing an index lookup, and it's really hard to tell that any of your operations preserve the order.

I suggest using tuples or named tuples or dictionaries to keep pairs of data that you want to perform a lookup over together.